[AMDGPU] V_SET_INACTIVE optimizations

perlfu opened the pull request:
Optimize V_SET_INACTIVE by allow it to run in WWM.
Hence WWM sections are not broken up for inactive lane setting.
WWM V_SET_INACTIVE can typically be lower to V_CNDMASK.
Some cases require use of exec manipulation V_MOV as previous code.
GFX9 sees slight instruction count increase in edge cases due to
smaller constant bus.

Additionally avoid introducing exec manipulation and V_MOVs where
a source of V_SET_INACTIVE is the destination.
This is a common pattern as WWM register pre-allocation often
assigns the same register.

perlfu committed:
[AMDGPU] V_SET_INACTIVE optimizations

Optimize V_SET_INACTIVE by always running it in run in WWM.
Allows WWM sections to be unbroken, and facilitates V_SET_INACTIVE
to be be lower to V_CNDMASK in most cases.
Some cases require use of exec manipulation V_MOV as previous code.
GFX9 sees slight instruction count increase in edge cases due to
smaller constant bus.

Additionally:
- Avoid introducing exec manipulation and V_MOVs where
  a source of V_SET_INACTIVE is the destination.
- Lower any V_SET_INACTIVE not touched by marking to COPY.

perlfu commented:
Currently includes #98858 as it touches many of the same tests.

perlfu commented:
- Rebase onto other WQM pass refactor
- Remove parts of #98562
- Remove semantics from V_SET_INACTIVE not used in WWM, always lower to COPY in these cases
- Optimize lowering of V_SET_INACTIVE to V_CNDMASK or V_MOV assuming WWM is available
- Update test IR where it appear to intend WWM behaviour for V_SET_INACTIVE but did not use WWM

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2270,37 +2270,162 @@ bool SIInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+    // present an issue.
+    // Fallback to V_MOV base lowering in all but the common cases.
+    const bool VMov64 = VMovOpc != AMDGPU::V_MOV_B32_e32;
+    const MachineFunction *MF = MI.getParent()->getParent();
```
```suggestion
    const MachineFunction *MF = MBB.getParent();
```

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2270,37 +2270,162 @@ bool SIInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+    unsigned VMovOpc = MI.getOpcode() == AMDGPU::V_SET_INACTIVE_B64
+                           ? AMDGPU::V_MOV_B64_PSEUDO
+                           : AMDGPU::V_MOV_B32_e32;
+    Register ExecReg = ST.isWave32() ? AMDGPU::EXEC_LO : AMDGPU::EXEC;
```
```suggestion
    Register ExecReg = RI.getExec();
```

perlfu committed:
- Address reviewer comments

perlfu commented:
- Rebase and squash
- Address outstanding comments

perlfu committed:
- Address reviewer comments.
- Add findImplicitExecSrc helper.
- Use helper to ignore V_SET_INACTIVE instructions during WQM/WWM
  processing.  This allows other passes to emit V_SET_INACTIVE
  for already known WWM sections.  This supports #105822.
- Add test for above.

perlfu commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2273,37 +2273,162 @@ bool SIInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+        (usesConstantBus(MRI, InactiveSrc, Desc.operands()[Src0Idx]) ? 1 : 0);
+    int LiteralConstants =
+        (ActiveSrc.isImm() && !isInlineConstant(ActiveImm) ? 1 : 0) +
+        (InactiveSrc.isImm() && !isInlineConstant(InactiveImm) ? 1 : 0);
```
Fixed

perlfu commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2273,37 +2273,162 @@ bool SIInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+        if (DstIsActive) {
+          MachineInstr *ExecMI =
+              BuildMI(MBB, MI, DL, get(NotOpc), ExecReg).addReg(ExecSrcReg);
+          ExecMI->addRegisterDead(AMDGPU::SCC, TRI); // SCC is overwritten
```
Done

perlfu commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2273,37 +2273,162 @@ bool SIInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+        continue;
+      ExecSrcReg = OpReg;
+      break;
+    }
```
Done

cdevadas commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2098,8 +2098,20 @@ unsigned SIInstrInfo::getNumWaitStates(const MachineInstr &MI) {
+      continue;
+    Register OpReg = Op.getReg();
+    if (OpReg == AMDGPU::EXEC || OpReg == AMDGPU::EXEC_LO ||
+        OpReg == AMDGPU::SCC)
```
The function name suggests it looks for implicit EXEC. Why is SCC here?
Also, it returns non-EXEC implicit Reg. The name is misleading.

perlfu replied:
```findSetInactiveMask```?

cdevadas replied:
Looks good.

perlfu committed:
findImplicitExecSrc -> findSetInactiveMask

jayfoad commented on llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.set.inactive.ll:
```diff
@@ -4,18 +4,39 @@
+; GCN-NEXT:    s_mov_b64 exec, s[4:5]
+; GCN-NEXT:    s_waitcnt lgkmcnt(0)
+; GCN-NEXT:    v_mov_b32_e32 v0, s6
+; GCN-NEXT:    s_mov_b64 exec, -1
+; GCN-NEXT:    s_mov_b64 exec, s[4:5]
```
This pair of instructions is redundant. It would be nice to remove them one day.

perlfu replied:
Agreed.

jayfoad commented on llvm/test/CodeGen/AMDGPU/global_atomics_scan_fadd.ll:
```diff
@@ -816,12 +816,10 @@ define amdgpu_kernel void @global_atomic_fadd_uni_address_div_value_agent_scope_
 ; GFX9-DPP-NEXT:    s_mov_b64 exec, s[0:1]
 ; GFX9-DPP-NEXT:    v_mov_b32_e32 v4, v0
-; GFX9-DPP-NEXT:    s_not_b64 exec, exec
-; GFX9-DPP-NEXT:    v_bfrev_b32_e32 v4, 1
-; GFX9-DPP-NEXT:    s_not_b64 exec, exec
-; GFX9-DPP-NEXT:    s_or_saveexec_b64 s[0:1], -1
+; GFX9-DPP-NEXT:    s_mov_b64 exec, -1
```
Even better would be to replace these four instructions with a single v_cndmask.

perlfu replied:
The GFX9 constant bus/literal limit prevents us merging all 4.
On GFX10+ they are all merged.
In this case we could use `v_cndmask_b32 v4, v0, v4, s[0:1]` for the last three as destination is one of the sources.

jayfoad commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.set.inactive.chain.arg.ll:
```diff
@@ -12,97 +12,204 @@ define amdgpu_cs_chain void @set_inactive_chain_arg(ptr addrspace(1) %out, i32 %
 ; GFX11-NEXT:    v_mov_b32_e32 v0, v10
-; GFX11-NEXT:    s_not_b32 exec_lo, exec_lo
-; GFX11-NEXT:    global_store_b32 v[8:9], v0, off
+; GFX11-NEXT:    s_mov_b32 exec_lo, s0
+; GFX11-NEXT:    s_or_saveexec_b32 s0, -1
```
This pair of instructions could be removed.

jayfoad reviewed (commented):
Just noting some places where the codegen could be improved in future.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -2273,37 +2287,159 @@ bool SIInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
+
+    if (UseVCndMask && VMov64) {
+      // Dual V_CNDMASK_B32
+      MachineOperand ActiveLo =
```
This feels like `SIInstrInfo::buildExtractSubRegOrImm` but I guess that assumes virtual registers and here you need physical registers?

arsenm replied:
It would be nicer to extract this into some variant of buildExtractSubRegOrImm that handles physical registers

perlfu replied:
I've made `buildExtractSubReg` do something for physical registers and used it instead for this case.

jayfoad commented on llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:
```diff
@@ -1058,6 +1071,20 @@ void SIWholeQuadMode::lowerBlock(MachineBasicBlock &MBB) {
+        MI.addOperand(*MBB.getParent(),
+                      MachineOperand::CreateReg(ActiveLanesReg, false, true));
+      } else
+        assert(State == StateExact || State == StateWQM);
```
Nit: braces around the else part, since the if part has them.

jayfoad reviewed (approved):
LGTM overall though I am not particularly confident reviewing the SIWholeQuadMode parts.

perlfu committed:
- Address reviewer comments

perlfu commented:
- Rebase for submission

jayfoad commented on llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:
```diff
@@ -555,16 +559,24 @@ char SIWholeQuadMode::scanInstructions(MachineFunction &MF,
-          } else {
-            markOperand(MI, Inactive, StateStrictWWM, Worklist);
+        // Ignore these if V_SET_INACTIVE which already has exec src register.
+        // These are generated by an earlier pass which has seperately ensured
+        // WWM and provided a mask of inactive lanes.
```
@perlfu what's the point of this? As far as I can tell it only happens with the hand written MIR test that you added (preloaded_set_inactive in wqm.mir).

perlfu replied:
This was added to facilitate `V_SET_INACTIVE` use in #105822.
